GSW/testUtility/testNS2.pyw

```diff
@@ -26,87 +26,148 @@ commandField = Entry(commandFrame, width=30, borderwidth=3, font=('Consolas',10)
+# send command #
+def sendCommand(event=None): # sends a command over serial
+    if not portOpen:
+        updateLog('No connection!')
+        return
     command = commandField.get()
     if command:
         teensy.write(bytes(command, 'utf-8'))
         updateLog('Command Sent: ' + command)
         commandField.delete(0, END)
-button_sendCommand = Button(commandFrame, text='Send', command=sendCommand)
-# bind the enter key to send command
-root.bind('<Return>', sendCommand)
 
-# read serial
-def readSerial():
+# read serial #
+def readSerial(): # reads incoming communication from NS2 via the open serial port
     data = str(teensy.readline())[2:-1]
     if data:
         monitor.configure(state ='normal')
-        monitor.insert(INSERT, codecs.decode(data, "unicode_escape"))
+        monitor.insert(END, codecs.decode(data, "unicode_escape")) # insert data in monitor
         monitor.configure(state ='disabled')
-        monitor.see(END)
+        if autoScroll.get():
+            monitor.see(END) 
 
 # scan for serial ports
-def scanPorts():
+def scanPorts(): # finds and lists all available com ports
     availablePorts = list_ports.comports()
     updateLog('==== Found Serial Ports: ====')
     for port in availablePorts:
         updateLog('- ' + port.description)
-button_scanPorts = Button(serialFrame, text='Scan For Serial Ports', command=scanPorts)
-        
-# update serial button
-def updateSerialPort():
-    teensy.port = portField.get()
-button_updatePort = Button(serialFrame, text='Set Serial Port', command=updateSerialPort)
+                
+# open serial port #
+def openSerialPort(): # opens a new serial connection if port exists
+    availablePorts = list_ports.comports()
+    for port in availablePorts: # if port exists, connect to it
+        if portField.get() in port.description:
+            teensy.close()
+            teensy.port = portField.get()
+            teensy.open()
+            global portOpen
+            portOpen = True
+            updateLog('Connected to port \"' + portField.get() + '\".')
+            return    
+    updateLog('Port \"' + portField.get() + '\" not found.')
+    
+# save to file #
+def saveFile(): # saves monitor contents to a file
+    fileName = fileNameField.get() + '.txt'
+    filePath = 'saved files\\' + fileNameField.get()
+    if (os.path.exists(filePath)): # display warning if file already exists
+        warning = '\"' + fileName + '\" already exists. Do you wish to overwrite it?'
+        overwrite = tkinter.messagebox.askyesno(title='Warning', message= warning)
+        if (not overwrite):
+            return
+    # write to file
+    file = open(filePath,'w', newline = '')
+    file.write(monitor.get('1.0','end-1c'))
+    updateLog( 'Output saved to ' + fileName)
+
+# clear output #
+def clearOutput(): # clears the output monitor
+    confirm = tkinter.messagebox.askyesno(title='Warning', message= 'Are you sure you want to clear the output monitor???')
```

Comment:
I wonder if a yes/no box would be good on certain commands as well.

Reply:
I added a flag in the FSW to enable/disable potentially dangerous commands that can cause shutdown or data loss. It will yell at you if you try to execute them without disabling the safeguard.